Fix #109 and #103

Akianonymus opened the pull request:
add a merge.sh script which will merge the scripts to form a single executable file

modify installtion script to only install the single file

because this is a drastic change in installation process, so add a special section in readme on how to remove older installtion

no new file will be created when global installtion is done, all the installtion related data will be stored in script itself

update contributing guide

update github action workflow to check if merge.sh was used

manage file perms, make everything readonly if possible
313ccd0

Akianonymus commented:
The ci checks are gonna fail because of the nature of pr

Added a section for new instructions on how migrate to new version

https://github.com/Akianonymus/google-drive-upload/tree/wip#migrate-from-old-version

labbots commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392&bid=20227261)

labbots commented on bash/upload.bash:
```diff
@@ -47,27 +47,29 @@ _short_help() {
-                _update 2>&1 1>| "${INFO_PATH}/update.log" &&
-                _update_config LAST_UPDATE_TIME "$(printf "%(%s)T\\n" "-1")" "${INFO_FILE}"
-        }
+        _REPO="${REPO}"
```
Codacy found an issue: [Possible misspelling: REPO may not be assigned, but _REPO is.](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392)

labbots commented on sh/upload.sh:
```diff
@@ -47,40 +47,40 @@ _short_help() {
-                _update 2>&1 1>| "${INFO_PATH}/update.log" &&
-                _update_config LAST_UPDATE_TIME "$(date +'%s')" "${INFO_FILE}"
-        }
+        _REPO="${REPO}"
```
Codacy found an issue: [Possible misspelling: REPO may not be assigned, but _REPO is.](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392)

labbots commented on install.sh:
```diff
@@ -565,23 +545,44 @@ _uninstall() {
         "${SYNC_COMMAND_NAME}" -k all 2>| /dev/null 1>&2 || :
-        rm -rf "${INFO_PATH}"/sync "${INSTALL_PATH:?}"/"${SYNC_COMMAND_NAME}"
+        chmod -f u+w "${INSTALL_PATH}/${SYNC_COMMAND_NAME}"
+        rm -rf "${INFO_PATH}"/sync "${INSTALL_PATH}/${SYNC_COMMAND_NAME}"
```
Codacy found an issue: [Use "${var:?}" to ensure this never expands to / .](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392)

Akianonymus commented:
@labbots any updates ?

labbots commented:
I ll look into this shortly

Akianonymus commented:
One major point after this pr, a clean installation will be needed for all users.

labbots commented:
@Akianonymus I am bit hesitant to roll out breaking changes. Moreover I dont see why it is necessary to merge all the scripts into one file and what advantage does it bring to the end user.

Akianonymus commented:
Yeah, same.

I will probably add some code to avoid breaking it, adding some logic to handle old existing installations.

About advantage, this will make it possible to have global installation.

Also, no need to inject utils file path to main script, just single file.

Not much advantage to the end user, but a necessary update.

But yeah, should be merged when old installations are handled properly.

labbots commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392&bid=20227261)

labbots commented on install.sh:
```diff
@@ -687,13 +715,19 @@ main() {
+                return 0
+            elif SCRIPT_VALUES="$(grep -E "${VALUES_REGEX}|^LATEST_INSTALLED_SHA=\".*\".* # added values|^SELF_SOURCE=\".*\"" "${COMMAND_PATH}" || :)" &&
+                eval "${SCRIPT_VALUES}" 2> /dev/null && [ -n "${LATEST_INSTALLED_SHA:+${SELF_SOURCE}}" ]; then
+                . "${CONFIG_INFO}"
```
Codacy found an issue: [Can't follow non-constant source. Use a directive to specify location.](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392)

labbots commented on install.sh:
```diff
@@ -687,13 +715,19 @@ main() {
-            if grep -q COMMAND_NAME "${INFO_PATH}"/google-drive-upload.info 2>| /dev/null 1>&2; then
+        if COMMAND_PATH="$(command -v "${COMMAND_NAME}")"; then
+            if [ -f "${INFO_PATH}/google-drive-upload.info" ] && [ -f "${INFO_PATH}/google-drive-upload.binpath" ] && [ -f "${INFO_PATH}/google-drive-upload.configpath" ]; then
+                OLD_INSTALLATION_PRESENT="true" && . "${INFO_PATH}/google-drive-upload.info"
```
Codacy found an issue: [Can't follow non-constant source. Use a directive to specify location.](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392)

Akianonymus commented:
So, i have pushed the required changes to avoid breaking existing installation and added additional code to cleanup while we update the user.

Added some strict checkings to detect old and new installation properly, to avoid any mishaps.

labbots commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392&bid=20227261)

labbots commented on bash/upload.bash:
```diff
@@ -93,16 +96,32 @@ _update() {
+_update_value() {
+    declare command_path="${INSTALL_PATH:?}/${COMMAND}" \
+        value_name="${1:-}" value="${2:-}" \
+        script_without_value_and_shebang="$(grep -v "${value_name}=\".*\".* # added values" "${command_path}" | sed 1d)"
```
Codacy found an issue: [Declare and assign separately to avoid masking return values.](https://app.codacy.com/manual/labbots/google-drive-upload/pullRequest?prid=6201392)

Akianonymus commented:
@labbots ?

Akianonymus commented:
Update: Needs more work

Akianonymus commented:
Do not use raw.githubusercontent.com urls, use redirect links

README: Add info about config files and how to use it effectively for automation

Akianonymus commented:
@labbots Please approve again.

Akianonymus commented:
A new release has been created: https://github.com/labbots/google-drive-upload/releases/tag/v3.4
